deepchem/models/callbacks.py

```diff
@@ -49,6 +48,10 @@ def __init__(self,
       if True, the best model is considered to be the one that minimizes the
       validation metric.  If False, the best model is considered to be the one
       that maximizes it.
+    logging_strategy: str
+      the logging strategy used for logging (step or epoch). If "step",
+      logging interval will be the value provided for `interval`. If "epoch",
+      then logging will happen at the end of every training epoch.
```

Comment:
I'm not sure this is exactly the behavior we want.  I suggest this flag determines whether `interval` is a number of steps or a number of epochs.  When training with a small dataset, you might only want to perform validation every 10 or 100 epochs.

Another option would be to deprecate the `interval` argument and replace it with two different arguments: `step_interval` and `epoch_interval`.  The user would specify one or the other, but not both.

Reply:
> After looking through the code some more, I suggest we separate out all the changes to logging behavior, like logging per epoch rather than per step. It's adding a lot of complexity and we need to think some more about the design. And it isn't really specific to Wandb. It will be simpler if we do a first PR that is just Wandb-specific changes while leaving the basic logging behavior the same, and then a separate PR with improvements to the behavior.
> 
> I notice a lot of formatting changes. We run all the code through yapf 0.22. If you use a different version, CI will complain about it. (Though we really should upgrade to a newer version one of these days.)

Ah my bad, I just saw this comment. Sure, we can remove all the changes for logging per epoch and also logging training metrics. I agree that it would be more complex and require more extensive modifications. We could do a simple PR first with just the current logging behaviour and go from there.